server/src/main/java/com/cloud/vm/schedule/VMScheduleManagerImpl.java

```diff
@@ -0,0 +1,295 @@
+                vmSchedule.setPeriod(period);
+            if (action != null)
+                vmSchedule.setAction(action);
+            if (tag != null)
+                vmSchedule.setTag(tag);
+            if (timezone != null)
+                vmSchedule.setTimezone(timezone);
+
+            vmScheduleDao.update(vmSchedule.getId(), vmSchedule);
+        } else {
+            throw new InvalidParameterValueException("Enable the state of VM Schedule before updating it");
+        }
+        return vmSchedule;
+    }
+
+    public static final class VMScheduleBackgroundTask extends ManagedContextRunnable implements BackgroundPollTask {
+        private  VMScheduleManagerImpl serviceImpl;
+
+        public VMScheduleBackgroundTask(VMScheduleManagerImpl serviceImpl) {
+            this.serviceImpl = serviceImpl;
+        }
+
+        private void scheduleActionOnVM(String action, VMSchedule.State state, String uuid, String period) {
+
+            serviceImpl.vmManager.start(uuid, null);
+        }
+
+        private Boolean matchPeriodWithCurrentTimestamp(String period){
+            Date currentTimestamp = new Date();
+            String[] periodParts = period.split(" ");
+
+
+            return true;
+        }
+
+        private void getVmSchedulesToBeExecuted() {
+
+        }
+
+        @Override
+        protected void runInContext() {
+            try {
+                if (s_logger.isTraceEnabled()) {
+                    s_logger.trace("VM Scheduler GC task is running...");
+                }
+                // Date currentTimestamp = new Date();
+                for (final VMScheduleVO vmSchedule :serviceImpl.vmScheduleDao.listAll()) {
+                    VMInstanceVO vmInstance = serviceImpl.vmInstanceDao.findById(vmSchedule.getVmId());
+                    scheduleActionOnVM(vmSchedule.getAction(), vmSchedule.getState(), vmInstance.getUuid(), vmSchedule.getPeriod());
+                }
+
+                /*
+                Long accountId= CallContext.current().getCallingAccountId();
+                List<VMInstanceVO> vmInstanceVOList = serviceImpl.vmInstanceDao.listByAccountId(accountId);
+                for (final VMInstanceVO vmInstance : vmInstanceVOList) {
+                    List<VMScheduleVO> vmScheduleVOList = serviceImpl.vmScheduleDao.findByVm(vmInstance.getId());
+                    if (vmScheduleVOList != null) {
+                        for (final VMScheduleVO vmSchedule : vmScheduleVOList) {
+                            scheduleActionOnVM(vmSchedule.getAction(), vmSchedule.getState(), vmInstance.getUuid(), vmSchedule.getPeriod());
+                        }
+                    }
+                }
+                */
+            } catch (final Throwable t) {
+                s_logger.error("Error trying to run VM Scheduler GC task", t);
+            }
+        }
+
+        @Override
+        public Long getDelay() {
+            // In Milliseconds
+            return VMSchedulerInterval.value() * 1000L;
```

Comment:
suggestion - need to check if you must multiply by 1000L (or if background poll manager does it already?)

Reply:
Sure

Reply:
Need to multiply by 1000L